90zfsbootmenu/lib/zfsbootmenu-kcl.sh

```diff
@@ -0,0 +1,258 @@
+    echo ""
+    return 0
+  fi
+
+  # KCL does not specify parent inheritance, just return the args
+  if ! [[ "${args}" =~ "%{parent}" ]]; then
+    zdebug "no parent reference in org.zfsbootmenu:commandline on ${zfsbe}"
+    echo "${args}"
+    return 0
+  fi
+
+  # Need to recursively expand "%{parent}"
+
+  parfs="${zfsbe%/*}"
+  if [ -z "${parfs}" ] || [ "${parfs}" = "${zfsbe}" ]; then
+    # There is no parent, par_args is empty
+    par_args=""
+  else
+    # Query the parent for KCL properties
+    if ! par_args="$( read_kcl_prop "${parfs}" )"; then
+      zwarn "failed to invoke read_kcl_prop on parent ${parfs}"
+      par_args=""
+    fi
+
+    # When the KCL property is inherited, recursive expansion fully populates
+    # the KCL at the level of the ancestor that actually defines the property.
+    if inherited="$( zfs get -H -o source -s inherited org.zfsbootmenu:commandline "${zfsbe}" 2>/dev/null )"; then
+      # Inherited property have a source of "inherited from <ancestor>";
+      # non-inherited properties will not be printed with `-s inherited`
+      if [ -n "${inherited}" ]; then
+        zdebug "org.zfsbootmenu:commandline on ${zfsbe} is inherited, using parent expansion verbatim"
+        echo "${par_args}"
+        return 0
+      fi
+    fi
+  fi
+
+  echo "${args//%\{parent\}/${par_args}}"
+  return 0
+}
+
+# arg1..argN: keys (and, optionally, associated value) to suppress from KCL
+# prints: tokenized KCL [read from stdin] with suppressed arguments removed
+
+kcl_suppress() {
+  local arg rem sup
+  while read -r arg; do
+    # Check match against all exclusions
+    sup=0
+    for rem in "$@"; do
+      # Arguments match entirely or up to first equal
+      if [ "${arg}" = "${rem}" ] || [ "${arg%%=*}" = "${rem}" ]; then
+        sup=1
+        break
+      fi
+    done
+
+    # Echo argument if it was not suppressed
+    [ "${sup}" -ne 1 ] && echo "${arg}"
+  done
+}
+
+# args1..argN: keys (and values, as appropriate) to append to a tokenized KCL
+# prints: tokenized KCL [read from stdin] with appended arguments
+
+kcl_append() {
+  local arg
+
+  # Carry forward input KCL
+  cat
+
+  # Append one line per argument
+  for arg in "$@"; do
+    echo "$arg"
+  done
+}
+
+
+# args: none
+# prints: space-separated concatenation of tokenized KCL [read from stdin]
+
+kcl_assemble() {
+  awk '
+    BEGIN{ strt = 1; }
+
+    {
+      if (strt == 0) {
+        printf " ";
+      }
+
+      printf "%s", $0;
+      strt = 0;
+    }
+  '
+}
+
+# args: none
+# prints: tokenized KCL from /etc/cmdline, /etc/cmdline.d/*.conf and /proc/cmdline
+
+get_zbm_kcl() {
+  local f
+
+  [ -r /etc/cmdline ] && kcl_tokenize < /etc/cmdline
+
+  for f in /etc/cmdline.d/*.conf; do
+    [ -r "${f}" ] || continue
+    kcl_tokenize < "${f}"
+  done
+
+  [ -r /proc/cmdline ] && kcl_tokenize < /proc/cmdline
+}
+
+# arg1..argN: argument to extract (multiples are alternatives, ordered by precedence)
```

Comment:
`getarg()` from Dracut had the idea of a deprecated option. Supporting `-d option` would add excessive complexity / bloat to `get_zbm_arg()`. Would it instead be reasonable/doable to log in some fashion (either zdebug or zinfo level) when a non-primary argument is matched? It'd be of marginal value for us, and essentially no value to end-users.

Reply:
Done. `get_zbm_arg` will now `zdebug` when the option that matches is not the first one specified.